sympy/printing/codeprinter.py

```diff
@@ -14,6 +16,32 @@
 from sympy.codegen.ast import Assignment
 
 
+def prints_statement(method):
+    """ Decorator for methods printing statements (as opposed to expressions) """
+
+    @wraps(method)
+    def wrapper(self, expr):
+        if self._settings.get('enable_statements', True):
+            return method(self, expr)
+        else:
+            raise ValueError("The code-printer setting 'enable_statements' is False,\n"
+                             "but %s generates a statement in the chosen language." %
+                             type(expr))
```

Comment:
When in doubt I think the printer should just assume that it is dealing with an expression (e.g `ccode(x**2)` shouldn't append a semicolon). An alternative would be to enable printing of statements in a subclass of the CodePrinter, but due to backward compatibility CodePrinter will need to continue to support printing statements.